compiler/rustc_infer/src/infer/error_reporting/mod.rs

```diff
@@ -1999,6 +1999,31 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
                         (ty::Bool, ty::Tuple(list)) => if list.len() == 0 {
                             self.suggest_let_for_letchains(&mut err, &trace.cause, span);
                         }
+                        // Suggest providing a std::ptr::null{,_mut}() if they pass in a 0 to an *{const,mut} uint
+                        (ty::RawPtr(ptr), ty::Uint(_))  => {
+                            if let Ok(code) = self.tcx.sess().source_map().span_to_snippet(span) {
+                                if let Ok(0) = code.parse::<u8>() {
+                                    match ptr.mutbl {
+                                        hir::Mutability::Not => {
+                                            err.span_suggestion(
+                                                span,
+                                                "if you meant to create a null pointer, use `std::ptr::null()`",
+                                                "std::ptr::null()",
```

Comment:
Hard-coding `std` may be annoying for `#![no_std]` crates, please get the path to the diagnostic item instead.

The only way I've found so far that does that is the following:
```rust
let null_sym = match ptr.mutbl {
    hir::Mutability::Not => sym::ptr_null,
    hir::Mutability::Mut => sym::ptr_null_mut,
};

if let Some(null_did) = self.tcx.get_diagnostic_item(null_sym) {
    if let Ok((printer, true)) = FmtPrinter::new(self.tcx, Namespace::ValueNS).try_print_visible_def_path(null_did) {
        let path = printer.into_buffer();
        err.span_suggestion(
            span,
            format!("if you meant to create a null pointer, use `{path}`"),
            path + "()",
            Applicability::MachineApplicable,
        );
    }
}
```

Reply:
all that `FmtPrinter` stuff might be avoided with a call to `with_no_trimmed_paths!(tcx.def_path_str(null_def_id))`

Reply:
Also, this logic may better live in `FnCtxt::emit_type_mismatch_suggestions` (well, a method called from it). At that point, we should have access to the HIR expression that causes this mismatch. Then we can just check if it's a literal. 

Otherwise, I would personally not want to land this if we're doing that span_to_snippet + parsing. It seems pretty fragile.

Reply:
Alternatively, we could probably make `report_and_explain_type_error` to take an optional `Option<&'tcx hir::Expr<'tcx>>`, and use that here, bubbling it down from hir typeck. That's a much more intrusive change, though.